src/Workspaces/Core/Portable/Workspace/Host/PersistentStorage/IPersistentStorageLocationService.cs

```diff
@@ -53,11 +21,18 @@ public DefaultPersistentStorageLocationService()
         {
         }
 
-        public bool IsSupported(Workspace workspace) => false;
+        public string TryGetStorageLocation(Solution solution)
+        {
+            if (string.IsNullOrWhiteSpace(solution.FilePath))
+                return null;
+
+            // Ensure that each unique workspace kind for any given solution has a unique
+            // folder to store their data in.
+            var checksums = new[] { Checksum.Create(solution.FilePath), Checksum.Create(solution.Workspace.Kind) };
+            var hashedName = Checksum.Create(WellKnownSynchronizationKind.Null, checksums).ToString();
+            var workingFolder = Path.Combine(Path.GetTempPath(), hashedName);
```

Comment:
this is the actual 'meat' of the change.  What we do is to get a unique folder location based on the pair of "solution file path" and "workspace kind".  This ensures, for example, that the main VS db gets a unique path distinct from the main Remote OOP db path.  i.e. they have the same solution path, but they have a different workspace kind.  

The use of checksum is simply to take these unique complex strings and form a nice unique base64 encoded folder name that will persist across sessions.

Reply:
Note: i considered even removing this "storage location" service and just hardcoding this logic into the storage-service.  However, it turns out the service is still useful. For example, this code doesn't work if the solution doesn't have a FilePath.  It turns out there are times this happens when someone might still want to use the persistence service.  For example, our tests and benchmark runners do this.  They often make adhoc workspaces and simply want to hardcode in a location where the sqlite db will be stored.  So i kept the service around accordingly.

Reply:
Not sure if it's good idea to store it in temp dir. Why not keep it in`.vs` dir, with the same hashed name?
I think keeping the cache local to the solution is better as it allows one to control cleanup of the cache per solution. For example, if I clean my temp dir I don't expect loading of all my solutions on the disk to get slower.

Reply:
Re locking of the file when VS is open - why do we need to keep the file open while VS is open? Can we not read the cached data and close the file immediately?

Reply:
It's odd to use _temp_ dir for _persistent_ storage (do you see the problem?).

Reply:
> Not sure if it's good idea to store it in temp dir. Why not keep it in.vs dir,

There are several problems with this:

First, vs can move this directory around.  This leads to an enormous amount of complexity (i.e. all the complexity reoved here) around tracking the directory, moving it around, ensuring OOP is in sync with it, etc.

Second, the directory gets locked, which leads to customer frustration. Note: this is not something we can efficiently lock/unlock as the cost at the sqlite level is far too high.  i.e. each time we we close open it sqlite has to go and recompute all the schema information.  It also means all our cached connections, sprocs, etc. are dumped and have to be recomputed.    I tried this out originally, and the cost was enormous (think 10-100x hit over being able to just keep pooled connections around and batch all our data through it through precompiled statements and whatnot).

Reply:
> It's odd to use temp dir for persistent storage (do you see the problem?).

I agree it seems slightly odd.  Another approach we could take would be to use the user's 'local persistence location'.  i.e. `%localappdata%` on windows.  The temp dir isn't actually much of a problem for us as it is something that is generally persistent on windows and because we're 100% resilient to this folder going away.  We only use it as an perf optimization, not a correctness one.  So as long as it stays around *most of the time* ,then we're good.  If it occasionally goes away, that's not an issue for us (just like we're fine if the user decides to wipe the .vs folder).

I'm open to alternative idea here.  However the price concern i have is getting rid of all the complexity of having to track and update the *vs specific* folder which might move around at runtime.  This is just pain for us and also leads to the locking issue.

Would you be ok with `%localappdata%/roslyn/sha` for windows and whatever would be equivalent on a unix system (probably `~/.roslyn`)?

Reply:
> Would you be ok with %localappdata%/roslyn/sha for windows and whatever would be equivalent on a unix system (probably ~/.roslyn)?

Note: i switched to this.  I like this a bit better as it is a clearly defined folder the user can delete as well to purge roslyn data.  Storing in `temp` means we read/write from a random location that we'd never be able to point them to to delete.

Reply:
FYI, this thread is getting long and hits several different topics.  If we continue this can we break out the separate topics into different messages so we can isolate the discussions? Thanks!

Reply:
Thinking about this a bit more I see the similarity with debugger symbol cache, which lives in `%temp%\SymbolCache` by default, which might not be the best choice. Every time I clean my temp dir I make my debugging slow because all symbols need to be redownloaded.  

`~\.roslyn\cache\sha` seems better to me. I'd add the `cache` subdir, in case we would like to use the dir for other data. It would be good to have UI in options to clean the cache.